webui/react/src/stores/cluster.tsx

```diff
@@ -0,0 +1,169 @@
+import { clone } from 'shared/utils/data';
+import { percent } from 'shared/utils/number';
+import { Agent, ClusterOverview, ClusterOverviewResource, ResourcePool, ResourceType } from 'types';
+import handleError from 'utils/error';
+import { Loadable, Loaded, NotLoaded } from 'utils/loadable';
+import { Observable, observable, useObservable, WritableObservable } from 'utils/observable';
+
+const initResourceTally: ClusterOverviewResource = { allocation: 0, available: 0, total: 0 };
+// TODO: dont export
+export const initClusterOverview: ClusterOverview = {
+  [ResourceType.CPU]: clone(initResourceTally),
+  [ResourceType.CUDA]: clone(initResourceTally),
+  [ResourceType.ROCM]: clone(initResourceTally),
+  [ResourceType.ALL]: clone(initResourceTally),
+  [ResourceType.UNSPECIFIED]: clone(initResourceTally),
+};
+
+class ClusterService {
+  agents: WritableObservable<Loadable<Agent[]>> = observable(NotLoaded);
+  resourcePools: WritableObservable<Loadable<ResourcePool[]>> = observable(NotLoaded);
+  pollingHandle: NodeJS.Timer;
+
+  canceler?: AbortController;
+
+  clusterOverview = this.agents.select((agents) =>
+    Loadable.map(agents, (agents) => {
+      const overview: ClusterOverview = clone(initClusterOverview);
+
+      agents.forEach((agent) => {
+        agent.resources
+          .filter((resource) => resource.enabled)
+          .forEach((resource) => {
+            const isResourceFree = resource.container == null;
+            const availableResource = isResourceFree ? 1 : 0;
+            overview[resource.type].available += availableResource;
+            overview[resource.type].total++;
+            overview[ResourceType.ALL].available += availableResource;
+            overview[ResourceType.ALL].total++;
+          });
+      });
+
+      for (const key in overview) {
+        const rt = key as ResourceType;
+        overview[rt].allocation =
+          overview[rt].total !== 0
+            ? percent((overview[rt].total - overview[rt].available) / overview[rt].total)
+            : 0;
+      }
+
+      return overview;
+    }),
+  );
+
+  clusterStatus = Observable.select(
+    [this.clusterOverview, this.resourcePools, this.agents],
+    (overview, pools, agents) => {
+      return Loadable.match(Loadable.all([overview, pools, agents]), {
+        Loaded: ([overview, pools, agents]) => clusterStatusText(overview, pools, agents) ?? '',
+        NotLoaded: () => undefined,
+      });
+    },
+  );
+
+  fetchAgents = async (signal: AbortSignal): Promise<void> => {
+    try {
+      const response = await getAgents({}, { signal });
+      this.agents.set(Loaded(response));
+    } catch (e) {
+      handleError(e);
+    }
+  };
+
+  fetchResourcePools = async (signal: AbortSignal): Promise<void> => {
+    try {
+      const response = await getResourcePools({}, { signal });
+      this.resourcePools.set(Loaded(response));
+    } catch (e) {
+      handleError(e);
+    }
+  };
+
+  poll = (): void => {
+    this.canceler?.abort();
```

Comment:
This appears to be redundant

Reply:
it cancels any in flight requests from the last round of polling, but yeah it doesn't really serve any purpose since the API call is not parameterized.

Reply:
This is the first line of `cancelPolling` though which you call the very next line. Cancelling the call is good since it guards against the case where the API is hanging.